dot/state/grandpa_changes.go

```diff
@@ -0,0 +1,313 @@
+
+// findApplicable try to retrieve an applicable change from the slice of forced changes
+func (oc orderedPendingChanges) findApplicable(importedHash common.Hash, importedNumber uint,
+	isDescendatOf isDescendantOfFunc) (*pendingChange, error) {
+
+	return oc.lookupChangeWhere(func(forced *pendingChange) (bool, error) {
+		announcingHash := forced.announcingHeader.Hash()
+		effectiveNumber := forced.effectiveNumber()
+
+		if importedHash.Equal(announcingHash) && effectiveNumber == importedNumber {
+			return true, nil
+		}
+
+		isDescendant, err := isDescendatOf(announcingHash, importedHash)
+		if err != nil {
+			return false, fmt.Errorf("cannot check ancestry: %w", err)
+		}
+
+		return isDescendant && effectiveNumber == importedNumber, nil
+	})
+
+}
+
+// lookupChangeWhere return the first pending change which satisfy the condition
+func (oc orderedPendingChanges) lookupChangeWhere(condition conditionFunc[*pendingChange]) (
+	pendingChange *pendingChange, err error) {
+	for _, change := range oc {
+		ok, err := condition(change)
+		if err != nil {
+			return nil, fmt.Errorf("failed while applying condition: %w", err)
+		}
+
+		if ok {
+			return change, nil
+		}
+	}
+
+	return nil, nil //nolint:nilnil
+}
+
+// importChange only tracks the pending change if and only if it is the
+// unique forced change in its fork, otherwise will return an error
+func (oc *orderedPendingChanges) importChange(pendingChange *pendingChange, isDescendantOf isDescendantOfFunc) error {
+	announcingHeader := pendingChange.announcingHeader.Hash()
+
+	for _, change := range *oc {
+		changeBlockHash := change.announcingHeader.Hash()
+
+		if changeBlockHash.Equal(announcingHeader) {
+			return fmt.Errorf("%w: %s", errDuplicateHashes, changeBlockHash)
+		}
+
+		isDescendant, err := isDescendantOf(changeBlockHash, announcingHeader)
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			return fmt.Errorf("%w: for block hash %s", errAlreadyHasForcedChange, changeBlockHash)
+		}
+	}
+
+	orderedChanges := *oc
+
+	// Use a binary search to include the pending change in the right position
+	// of a slice ordered by the effective number and by announcing header number
+	idxToInsert := sort.Search(oc.Len(), func(i int) bool {
+		return orderedChanges[i].effectiveNumber() >= pendingChange.effectiveNumber() &&
+			orderedChanges[i].announcingHeader.Number >= pendingChange.announcingHeader.Number
+	})
+
+	orderedChanges = append(orderedChanges, pendingChange)
+	copy(orderedChanges[idxToInsert+1:], orderedChanges[idxToInsert:])
+	orderedChanges[idxToInsert] = pendingChange
+	*oc = orderedChanges
+
+	return nil
+}
+
+// pruneChanges will remove changes whose are not descendant of the hash argument
+// this function updates the current state of the change tree
+func (oc *orderedPendingChanges) pruneChanges(hash common.Hash, isDescendantOf isDescendantOfFunc) error {
+	onBranchForcedChanges := make([]*pendingChange, 0, oc.Len())
```

Comment:
Using `append` on slice with a defined capacity will double the capacity on overflow.

Reply:
Yes, but in this case, the maximum amount of data in the `onBranchForcedChanges` slice is `oc.Len()`, so there is no problem to use append as it will never overflow the capacity